servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpPayloadDiscardCleanerServiceFilter.java

```diff
@@ -0,0 +1,131 @@
+
+import static io.servicetalk.http.netty.HttpPayloadDiscardWatchdogServiceFilter.payloadPublisherKey;
+import static io.servicetalk.http.netty.HttpPayloadDiscardWatchdogServiceFilter.payloadSubscribedKey;
+
+final class HttpPayloadDiscardCleanerServiceFilter implements StreamingHttpServiceFilterFactory, HttpLifecycleObserver {
+
+    /**
+     * Instance of {@link HttpPayloadDiscardCleanerServiceFilter}.
+     */
+    public static final StreamingHttpServiceFilterFactory INSTANCE = new HttpPayloadDiscardCleanerServiceFilter();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpPayloadDiscardCleanerServiceFilter.class);
+
+    private HttpPayloadDiscardCleanerServiceFilter() {
+        // Singleton
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(final StreamingHttpService service) {
+
+        final WatchdogHttpLifecycleObserver observer = new WatchdogHttpLifecycleObserver(this, false);
+
+        return new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return observer.trackLifecycle(ctx, request, r -> delegate().handle(ctx, r, responseFactory));
+            }
+
+            @Override
+            public HttpExecutionStrategy requiredOffloads() {
+                return HttpExecutionStrategies.offloadNone();
+            }
+        };
+    }
+
+    private static class WatchdogHttpLifecycleObserver extends AbstractLifecycleObserverHttpFilter {
+        WatchdogHttpLifecycleObserver(final HttpLifecycleObserver observer, final boolean client) {
+            super(observer, client);
+        }
+    }
+
+    @Override
+    public HttpExchangeObserver onNewExchange() {
+
+        return new HttpExchangeObserver() {
+
+            @Nullable
+            private HttpRequestMetaData requestMetaData;
+
+            @Nullable
+            private HttpResponseMetaData responseMetaData;
+
+            @Override
+            public HttpRequestObserver onRequest(final HttpRequestMetaData requestMetaData) {
+                this.requestMetaData = requestMetaData;
+                return NoopHttpLifecycleObserver.NoopHttpRequestObserver.INSTANCE;
+            }
+
+            @Override
+            public HttpResponseObserver onResponse(final HttpResponseMetaData responseMetaData) {
+                this.responseMetaData = responseMetaData;
+                return NoopHttpLifecycleObserver.NoopHttpResponseObserver.INSTANCE;
+            }
+
+            @Override
+            public void onExchangeFinally() {
+                if (requestMetaData != null && responseMetaData != null) {
+                    Boolean subscribed = requestMetaData.context().get(payloadSubscribedKey);
```

Comment:
if a retry occurs the request context map state will be shared, do we reset this state anywhere?

Reply:
ah server side retries are less likely but in principal could be done by inserting a filter that has a fallback to in-memory publisher if failure occurs.

Reply:
Added a test in the latest changeset.